Build failure on Debian 9 with boost 1.62

When trying to build the latest version of nmos-cpp with the latest cpprestsdk on Debian 9 (with the system Boost libraries V1.62), I get three errors in client_utils.cpp:

[ 83%] Building CXX object CMakeFiles/nmos-cpp_static.dir/nmos/components.cpp.o
/home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp: In instantiation of 'nmos::details::make_client_ssl_context_callback(const settings&)::<lambda(boost::asio::ssl::context&)> [with ExceptionType = web::http::http_exception]':
/home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp:32:21:   required from 'struct nmos::details::make_client_ssl_context_callback(const settings&) [with ExceptionType = web::http::http_exception; nmos::settings = web::json::value]::<lambda(class boost::asio::ssl::context&)>'
/home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp:44:13:   required from 'std::function<void(boost::asio::ssl::context&)> nmos::details::make_client_ssl_context_callback(const settings&) [with ExceptionType = web::http::http_exception; nmos::settings = web::json::value]'
/home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp:58:118:   required from here
/home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp:42:27: error: invalid user-defined conversion from 'const boost::system::error_code' to 'int' [-fpermissive]
                     throw ExceptionType(e.code(), e.what());
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/boost/asio/io_service.hpp:25:0,
                 from /usr/include/boost/asio/ssl/context.hpp:26,
                 from /home/michael/packages/nmos-cpp/Development/boost/asio/ssl/set_cipher_list.hpp:8,
                 from /home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp:4:
/usr/include/boost/system/error_code.hpp:362:7: note: candidate is: boost::system::error_code::operator boost::system::error_code::unspecified_bool_type() const <near match>
       operator unspecified_bool_type() const  BOOST_SYSTEM_NOEXCEPT // true if error
       ^~~~~~~~
/usr/include/boost/system/error_code.hpp:362:7: note:   no known conversion from 'boost::system::error_code::unspecified_bool_type {aka void (*)()}' to 'int'
In file included from /usr/local/include/cpprest/http_client.h:50:0,
                 from /home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp:7:
/usr/local/include/cpprest/http_msg.h:222:5: note:   initializing argument 1 of 'web::http::http_exception::http_exception(int, const string_t&)'
     http_exception(int errorCode, const utility::string_t& whatArg)
     ^~~~~~~~~~~~~~
/home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp: In instantiation of 'nmos::details::make_client_ssl_context_callback(const settings&)::<lambda(boost::asio::ssl::context&)> [with ExceptionType = web::websockets::client::websocket_exception]':
/home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp:32:21:   required from 'struct nmos::details::make_client_ssl_context_callback(const settings&) [with ExceptionType = web::websockets::client::websocket_exception; nmos::settings = web::json::value]::<lambda(class boost::asio::ssl::context&)>'
/home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp:44:13:   required from 'std::function<void(boost::asio::ssl::context&)> nmos::details::make_client_ssl_context_callback(const settings&) [with ExceptionType = web::websockets::client::websocket_exception; nmos::settings = web::json::value]'
/home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp:73:137:   required from here
/home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp:42:27: error: invalid user-defined conversion from 'const boost::system::error_code' to 'int' [-fpermissive]
                     throw ExceptionType(e.code(), e.what());
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/boost/asio/io_service.hpp:25:0,
                 from /usr/include/boost/asio/ssl/context.hpp:26,
                 from /home/michael/packages/nmos-cpp/Development/boost/asio/ssl/set_cipher_list.hpp:8,
                 from /home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp:4:
/usr/include/boost/system/error_code.hpp:362:7: note: candidate is: boost::system::error_code::operator boost::system::error_code::unspecified_bool_type() const <near match>
       operator unspecified_bool_type() const  BOOST_SYSTEM_NOEXCEPT // true if error
       ^~~~~~~~
/usr/include/boost/system/error_code.hpp:362:7: note:   no known conversion from 'boost::system::error_code::unspecified_bool_type {aka void (*)()}' to 'int'
In file included from /home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp:8:0:
/usr/local/include/cpprest/ws_client.h:249:5: note:   initializing argument 1 of 'web::websockets::client::websocket_exception::websocket_exception(int, const string_t&)'
     websocket_exception(int errorCode, const utility::string_t& whatArg)
     ^~~~~~~~~~~~~~~~~~~
[ 83%] Building CXX object CMakeFiles/nmos-cpp_static.dir/nmos/connection_activation.cpp.o
[ 84%] Building CXX object CMakeFiles/nmos-cpp_static.dir/nmos/connection_api.cpp.o
CMakeFiles/nmos-cpp_static.dir/build.make:166: recipe for target 'CMakeFiles/nmos-cpp_static.dir/cpprest/ws_listener_impl.cpp.o' failed
make[2]: *** [CMakeFiles/nmos-cpp_static.dir/cpprest/ws_listener_impl.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /usr/include/c++/6/mutex:42:0,
                 from /usr/local/include/pplx/pplxlinux.h:33,
                 from /usr/local/include/pplx/pplx.h:49,
                 from /usr/local/include/pplx/pplxtasks.h:61,
                 from /usr/local/include/cpprest/asyncrt_utils.h:17,
                 from /usr/local/include/cpprest/json.h:18,
                 from /home/michael/packages/nmos-cpp/Development/cpprest/json_ops.h:4,
                 from /home/michael/packages/nmos-cpp/Development/cpprest/json_utils.h:4,
                 from /home/michael/packages/nmos-cpp/Development/nmos/settings.h:4,
                 from /home/michael/packages/nmos-cpp/Development/nmos/client_utils.h:4,
                 from /home/michael/packages/nmos-cpp/Development/nmos/client_utils.cpp:1:
/usr/include/c++/6/functional:2106:7: error: 'std::function<_Res(_ArgTypes ...)>::function(_Functor) [with _Functor = nmos::details::make_client_ssl_context_callback(const settings&) [with ExceptionType = web::http::http_exception; nmos::settings = web::json::value]::<lambda(boost::asio::ssl::context&)>; <template-parameter-2-2> = void; <template-parameter-2-3> = void; _Res = void; _ArgTypes = {boost::asio::ssl::context&}]', declared using local type 'nmos::details::make_client_ssl_context_callback(const settings&) [with ExceptionType = web::http::http_exception; nmos::settings = web::json::value]::<lambda(boost::asio::ssl::context&)>', is used but never defined [-fpermissive]
       function<_Res(_ArgTypes...)>::
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/6/functional:2106:7: error: 'std::function<_Res(_ArgTypes ...)>::function(_Functor) [with _Functor = nmos::details::make_client_ssl_context_callback(const settings&) [with ExceptionType = web::websockets::client::websocket_exception; nmos::settings = web::json::value]::<lambda(boost::asio::ssl::context&)>; <template-parameter-2-2> = void; <template-parameter-2-3> = void; _Res = void; _ArgTypes = {boost::asio::ssl::context&}]', declared using local type 'nmos::details::make_client_ssl_context_callback(const settings&) [with ExceptionType = web::websockets::client::websocket_exception; nmos::settings = web::json::value]::<lambda(boost::asio::ssl::context&)>', is used but never defined [-fpermissive]
CMakeFiles/nmos-cpp_static.dir/build.make:218: recipe for target 'CMakeFiles/nmos-cpp_static.dir/nmos/client_utils.cpp.o' failed
make[2]: *** [CMakeFiles/nmos-cpp_static.dir/nmos/client_utils.cpp.o] Error 1
CMakeFiles/Makefile2:261: recipe for target 'CMakeFiles/nmos-cpp_static.dir/all' failed
make[1]: *** [CMakeFiles/nmos-cpp_static.dir/all] Error 2
Makefile:140: recipe for target 'all' failed

It seems to suggest adding'-fpermissive' in the right place may help me, but I'm not 100% sure where the right place would be, or whether there is a better fix.

--- a/Development/cmake/NmosCppLibraries.cmake
+++ b/Development/cmake/NmosCppLibraries.cmake
@@ -559,6 +559,7 @@ set(NMOS_CPP_CPPREST_DETAILS_HEADERS
     ${NMOS_CPP_DIR}/cpprest/details/boost_u_workaround.h
     ${NMOS_CPP_DIR}/cpprest/details/pop_u.h
     ${NMOS_CPP_DIR}/cpprest/details/push_undef_u.h
+    ${NMOS_CPP_DIR}/cpprest/details/system_error.h
     )
 
 set(NMOS_CPP_NMOS_SOURCES

--- a/Development/cpprest/details/system_error.h
+++ b/Development/cpprest/details/system_error.h
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <map>
+#include <memory>
 #include <mutex>
 #include <system_error>
 #include <boost/system/system_error.hpp>

--- a/Development/cpprest/details/system_error.h
+++ b/Development/cpprest/details/system_error.h
@@ -38,7 +38,8 @@ namespace web
             if (boost::system::generic_category() == category) return std::generic_category();
 
             struct cat_map_less { bool operator()(const boost::system::error_category* lhs, const boost::system::error_category* rhs) const { return *lhs < *rhs; } };
-            static std::map<const boost::system::error_category*, std::unique_ptr<boost_system_error_category>, cat_map_less> cat_map;
+            typedef std::map<const boost::system::error_category*, std::unique_ptr<boost_system_error_category>, cat_map_less> cat_map_type;
+            static cat_map_type cat_map;
             static std::mutex cat_mutex;
 
             std::lock_guard<std::mutex> guard(cat_mutex);
@@ -46,7 +47,7 @@ namespace web
             auto cat = cat_map.find(&category);
             if (cat_map.end() == cat)
             {
-                cat = cat_map.insert({ &category, std::unique_ptr<boost_system_error_category>(new boost_system_error_category(&category)) }).first;
+                cat = cat_map.insert(cat_map_type::value_type{ &category, std::unique_ptr<boost_system_error_category>(new boost_system_error_category(&category)) }).first;
             }
 
             return *cat->second;

--- a/Development/cpprest/details/system_error.h
+++ b/Development/cpprest/details/system_error.h
@@ -1,0 +1,78 @@
+#pragma once
+
+#include <map>
+#include <mutex>
+#include <system_error>
+#include <boost/system/system_error.hpp>
+
+namespace web
+{
+    namespace details
+    {
+#if BOOST_VERSION >= 106500
+        inline const std::error_category& to_std_error_category(const boost::system::error_category& category)
+        {
+            return category;
+        }
+#else
+        inline const std::error_category& to_std_error_category(const boost::system::error_category& category);
+
+        class boost_system_error_category : public std::error_category
+        {
+        private:
+            const boost::system::error_category* impl;
+
+        public:
+            explicit boost_system_error_category(const boost::system::error_category* impl) : impl(impl) {}
+
+            virtual const char* name() const BOOST_NOEXCEPT { return impl->name(); }
+            virtual std::error_condition default_error_condition(int code) const BOOST_NOEXCEPT { auto condition = impl->default_error_condition(code); return{ condition.value(), to_std_error_category(condition.category()) }; }
+            virtual bool equivalent(int code, const std::error_condition& condition) const BOOST_NOEXCEPT { return default_error_condition(code) == condition; }
+            virtual bool equivalent(const std::error_code& code, int condition) const BOOST_NOEXCEPT { return *this == code.category() && code.value() == condition; }
+            virtual std::string message(int condition) const { return impl->message(condition); }
+        };
+
+        inline const std::error_category& to_std_error_category(const boost::system::error_category& category)
+        {
+            if (boost::system::system_category() == category) return std::system_category();
+            if (boost::system::generic_category() == category) return std::generic_category();
+
+            struct cat_map_less { bool operator()(const boost::system::error_category* lhs, const boost::system::error_category* rhs) const { return *lhs < *rhs; } };
+            static std::map<const boost::system::error_category*, std::unique_ptr<boost_system_error_category>, cat_map_less> cat_map;
+            static std::mutex cat_mutex;
+
+            std::lock_guard<std::mutex> guard(cat_mutex);
+
+            auto cat = cat_map.find(&category);
+            if (cat_map.end() == cat)
+            {
+                cat = cat_map.insert({ &category, std::unique_ptr<boost_system_error_category>(new boost_system_error_category(&category)) }).first;
+            }
+
+            return *cat->second;
+        }
+#endif
+
+        inline std::error_code to_std_error_code(const boost::system::error_code& ec)
+        {
+            return{ ec.value(), to_std_error_category(ec.category()) };
+        }
+
+        inline bool equal_to(const boost::system::error_code& lhs, const boost::system::error_code& rhs) { return lhs == rhs; }
+        inline bool equal_to(const boost::system::error_code& lhs, const std::error_code& rhs) { return to_std_error_code(lhs) == rhs; }
+        inline bool equal_to(const std::error_code& lhs, const boost::system::error_code& rhs) { return lhs == to_std_error_code(rhs); }
+        inline bool equal_to(const std::error_code& lhs, const std::error_code& rhs) { return lhs == rhs; }
+
+        template <typename ExceptionType>
+        inline ExceptionType from_boost_system_system_error(const boost::system::system_error& e)
+        {
+            return ExceptionType(to_std_error_code(e.code()), e.what());
+        }
+
+        template <>
+        inline boost::system::system_error from_boost_system_system_error<boost::system::system_error>(const boost::system::system_error& e)
+        {
+            return e;
+        }
+    }
+}

--- a/Development/cpprest/ws_listener_impl.cpp
+++ b/Development/cpprest/ws_listener_impl.cpp
@@ -21,6 +21,7 @@ __pragma(warning(disable:4701)) // e.g. potentially uninitialized local variable
 PRAGMA_WARNING_POP
 
 #include "cpprest/asyncrt_utils.h" // for utility::conversions
+#include "cpprest/details/system_error.h"
 #include "cpprest/uri_schemes.h"
 
 // websocket_listener is an experimental server-side implementation of WebSockets
@@ -324,7 +325,7 @@ namespace web
                                 // if the error is "Underlying Transport Error" (pass_through), this might be a platform that doesn't support IPv6
                                 // (and depending on configuration, one can't detect boost::asio::error::address_family_not_supported directly)
                                 // since WebSocket++ 0.8.0, the error category and code used in this case seem to have changed
-                                if (make_error_code(boost::asio::error::address_family_not_supported) == ec
+                                if (web::details::equal_to(make_error_code(boost::asio::error::address_family_not_supported), ec)
                                     || make_error_code(websocketpp::transport::asio::error::pass_through) == ec
                                     || make_error_code(websocketpp::transport::error::pass_through) == ec)
                                 {

--- a/Development/nmos/client_utils.cpp
+++ b/Development/nmos/client_utils.cpp
@@ -4,6 +4,7 @@
 #include "boost/asio/ssl/set_cipher_list.hpp"
 #endif
 #include "cpprest/basic_utils.h"
+#include "cpprest/details/system_error.h"
 #include "cpprest/http_client.h"
 #include "cpprest/ws_client.h"
 #include "nmos/ssl_context_options.h"
@@ -39,7 +40,7 @@ namespace nmos
                 }
                 catch (const boost::system::system_error& e)
                 {
-                    throw ExceptionType(e.code(), e.what());
+                    throw web::details::from_boost_system_system_error<ExceptionType>(e);
                 }
             };
         }

--- a/Development/nmos/server_utils.cpp
+++ b/Development/nmos/server_utils.cpp
@@ -6,6 +6,7 @@
 #include "boost/asio/ssl/use_tmp_ecdh.hpp"
 #endif
 #include "cpprest/basic_utils.h"
+#include "cpprest/details/system_error.h"
 #include "cpprest/http_listener.h"
 #include "cpprest/ws_listener.h"
 #include "nmos/ssl_context_options.h"
@@ -30,7 +31,7 @@ namespace nmos
 
                     if (private_key_files.size() == 0)
                     {
-                        throw ExceptionType({},"Missing private key file");
+                        throw ExceptionType({}, "Missing private key file");
                     }
                     for (const auto& private_key_file : private_key_files.as_array())
                     {
@@ -39,7 +40,7 @@ namespace nmos
 
                     if (certificate_chain_files.size() == 0)
                     {
-                        throw ExceptionType({},"Missing certificate chain file");
+                        throw ExceptionType({}, "Missing certificate chain file");
                     }
                     for (const auto& certificate_chain_file : certificate_chain_files.as_array())
                     {
@@ -55,7 +56,7 @@ namespace nmos
                 }
                 catch (const boost::system::system_error& e)
                 {
-                    throw ExceptionType(e.code(), e.what());
+                    throw web::details::from_boost_system_system_error<ExceptionType>(e);
                 }
             };
         }

